### PR TITLE
[hot_reload] implement param reflection

### LINKED ISSUE #77512: [mono][hot reload] MethodBase.GetParameters on added method causes a crash
### Description

Seen when adding a new event handler with in an Android MAUI app  after adding a new XAML button.

Crash stack trace: https://gist.github.com/lambdageek/7055193d9078cbe48fce72f0d3675b3c

### Reproduction Steps

1. Create a new MAUI app and run it with debugging on Android (probably ios, too).
2. Add a new `<Button>` to the XAML with the tab-completed `"Button_Clicked"` event handler
3. Wait for the XAML change to apply
4. hit the "Apply Changes" VS button to apply the C# changes
5. Crash `* Assertion at /__w/1/s/src/mono/mono/metadata/metadata.c:1306, condition 'idx >= 0' not met`

### Expected behavior

Change is applied and the new event handler is usable

### Actual behavior

We hit an assertion

### Regression?

No

### Known Workarounds

restart the app

### Configuration

.NET 7 RC2
VS 17.4 Preview

### Other information

From the stack trace, it's crashing when we're trying to call `MethodBase.GetParameters`. the param field on the added method is 0. (method params are added with the AddParam EnC function code, and the param field on the MethodDef is 0)

Related AzDo ticket https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1567043

### PR DESCRIPTION
**Pull request by lambdageek**
Fixes https://github.com/dotnet/runtime/issues/77512

When a method is added, the paramlist column in the MethodDef is empty.  Instead, the parameter info is added by separate EnCLog "Add Param" commands.  Store that info in the image, keyed on the method definition token.

### CODE REVIEW

**Comment by lambdageek** [comment]
I would like to backport this for 7.0.x servicing

**Review by vargaz — APPROVED** [review]

File: src/libraries/System.Runtime.Loader/tests/ApplyUpdateTest.cs [hunk 1]
```diff
@@ -629,5 +629,52 @@ public static void TestReflectionAddNewType()
...
+
+		var parms = mi.GetParameters();
+
+		foreach (var parm in parms) {
+		    Console.WriteLine ($"parm: {parm}");
+		    var cas = parm.GetCustomAttributes();
+		    foreach (var ca in cas) {
+			Console.WriteLine ($" - ca: {ca}");
+		    }
+		}
```
Comment (lambdageek) [inline src/libraries/System.Runtime.Loader/tests/ApplyUpdateTest.cs:674]: oops. I didn't mean to leve these WriteLines in here. Need to convert them into asserts.

File: src/mono/mono/component/hot_reload-internals.h [hunk 2]
```diff
@@ -86,4 +86,10 @@ typedef struct _MonoClassMetadataUpdateEvent {
 	uint32_t token; /* the Event table token where this event was defined. */
 } MonoClassMetadataUpdateEvent;
 
+typedef struct _MonoMethodMetadataUpdateParamInfo {
+	uint32_t method_token; /* which method is this about */
```
Comment (lambdageek) [inline src/mono/mono/component/hot_reload-internals.h:90]:
```suggestion
```

File: src/mono/mono/component/hot_reload.c [hunk 3]
```diff
@@ -2366,9 +2389,26 @@ apply_enclog_pass2 (Pass2Context *ctx, MonoImage *image_base, BaselineInfo *base
...
-			 * FIXME: we need a lookaside table (like member_parent) for every place
-			 * that looks at MONO_METHOD_PARAMLIST
 			 */
+			if (is_addition) {
+				g_assert (add_field_method != 0);
+				/* 			
+				 * FIXME: we need a lookaside table (like member_parent) for every place
+				 * that looks at MONO_METHOD_PARAMLIST
+				 */
+
```
Comment (lambdageek) [inline src/mono/mono/component/hot_reload.c:2399]:
```suggestion
```

File: src/mono/mono/metadata/custom-attrs.c [hunk 4]
```diff
@@ -2250,12 +2250,20 @@ mono_custom_attrs_from_param_checked (MonoMethod *method, guint32 param, MonoErr
 		return NULL;
 	ca = &image->tables [MONO_TABLE_METHOD];
 
-	/* FIXME: metadata-update */
+	/* FIXME: metadata-update - lookup added params */
```
Comment (lambdageek) [inline src/mono/mono/metadata/custom-attrs.c:2253]:
```suggestion
```

**Comment by lambdageek** [comment]
@vargaz mind taking another look? same approach as before, but I put all the logic about figuring out the first & last param of a method into a separate function.

**Pull request merged**